src/driver/hermes_mod.c

```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
 };
 MODULE_DEVICE_TABLE(pci, pci_ids);
 
+static void irq_teardown(struct xdma_dev *xdev)
+{
+	irq_msix_channel_teardown(xdev);
+}
+
+static int irq_setup(struct xdma_dev *xdev)
+{
+	return xdma_irq_setup(xdev);
+}
+
+/*
+ * code to detect if MSI/MSI-X capability exists is derived
+ * from linux/pci/msi.c - pci_msi_check_device
+ */
+
+#ifndef arch_msi_check_device
+static int arch_msi_check_device(struct pci_dev *dev, int nvec, int type)
+{
+	return 0;
+}
+#endif
+
+/* type = PCI_CAP_ID_MSI or PCI_CAP_ID_MSIX */
+static int msi_msix_capable(struct pci_dev *dev, int type)
+{
+	struct pci_bus *bus;
+	int ret;
+
+	if (!dev || dev->no_msi)
+		return 0;
+
+	for (bus = dev->bus; bus; bus = bus->parent)
+		if (bus->bus_flags & PCI_BUS_FLAGS_NO_MSI)
+			return 0;
+
+	ret = arch_msi_check_device(dev, 1, type);
+	if (ret)
+		return 0;
+
+	if (!pci_find_capability(dev, type))
+		return 0;
+
+	return 1;
+}
+
+static void disable_msi_msix(struct pci_dev *pdev)
+{
+	pci_disable_msix(pdev);
+}
+
+static int enable_msi_msix(struct xdma_dev *xdev)
+{
+	struct pci_dev *pdev = xdev->pdev;
+	int rv = 0;
+
+	if (unlikely(!xdev || !pdev)) {
+		pr_err("xdev 0x%p, pdev 0x%p.\n", xdev, pdev);
+		return -EINVAL;
+	}
+
+	if (msi_msix_capable(pdev, PCI_CAP_ID_MSIX)) {
+		int req_nvec = xdev->c2h_channel_max + xdev->h2c_channel_max;
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
+		pr_debug("Enabling MSI-X\n");
+		rv = pci_alloc_irq_vectors(pdev, req_nvec, req_nvec,
```

Comment:
I'm also not sure I'm following exactly what's going on with all these moves and renames (bit of a rats nest).... But pci_alloc_irq_vectors() really should only be called by the PCI driver exactly once. It's not clear if that is still true here...

Reply:
Yeah, the main motivation for this PR was that the call to pci_alloc_irq_vectors() was deep inside the xdma init and we needed to add the extra vectors for eBPF. The new code calls it only once.